### ISSUE TITLE
Fix S3655 FP: HasValue checked with Pattern matching

### ISSUE THREAD

**Issue by pavel-mikula-sonarsource** [SonarSource/sonar-dotnet#6682]
Reported here: https://community.sonarsource.com/t/rule-c-s3655-false-positive-for-pattern/81703

```
public struct Repro_6682
{
    public bool SomeProperty { get; }

    public void PatternMatching(Repro_6682? arg, bool condition)
    {
        if (condition)
        {
            arg = null;
        }

        if (arg is { SomeProperty: true })   // A null check
        {
            var value = arg.Value;  // Noncompliant FP
        }

        if (arg is { })   // A null check
        {
            var value = arg.Value;  // Noncompliant FP
        }
    }
}
```

**Comment by Gakk** [comment]
Thanks for confirming this issue @pavel-mikula-sonarsource 👍

You gave a timeframe for Q1 [on the communit](https://community.sonarsource.com/t/rule-c-s3655-false-positive-for-pattern/81703)y:
> We plan to rework this rule in Q1 and that should also fix this issue.

Currently our developers get **failed** on the quality gate:
![image](https://user-images.githubusercontent.com/5013909/215074957-1d67b8fe-5640-44fd-9c19-a800dd7a99cd.png)

Is there any way to modify this rule using SonarCloud, without lowering the quality gate for all rules?

**Comment by pavel-mikula-sonarsource** [comment]
You should mark it as False Positive in SonarCloud UI

**Comment by Gakk** [comment]
> You should mark it as False Positive in SonarCloud UI

I thought marking in SonarCloud UI also involved manually overriding check on GitHub pull request (or re-run analyze), but your integration updated both comment and check immediately! Kudos! 🥳🚀

**Comment by martin-strecker-sonarsource** [comment]
Fixed as part of spring 8.56
